### ISSUE TITLE
[bug][tgi][atari_lynx] TGI_COLOR_BLACK is transparent and not a solid color

### ISSUE THREAD

**Issue by Fabrizio-Caruso** [cc65/cc65#1584]
Issue #956 was never fully solved.
On a white background, it is not possible to display black text.
The TGI_COLOR_BLACK is still implemented as transparent. 
TGI_COLOR_BLACK should be a real solid color.

The problem is reproduced with this example:
```c
#include <tgi.h>
#define CLI() asm("\tcli")
int main(void)
{
    tgi_install (tgi_static_stddrv);
    tgi_init ();        
    CLI();
    while (tgi_busy())  {  };
    tgi_setpalette(tgi_getdefpalette());
    tgi_setbgcolor(TGI_COLOR_WHITE);
    tgi_outtextxy(8,8,"    "); 
    tgi_setcolor(TGI_COLOR_BLACK);
    // tgi_setcolor(TGI_COLOR_RED);
    tgi_outtextxy(8,8,"hello");
        
    while(1){};
    return 0;
}
```
@greg-king5 We still have a problem with the black color... Do you know what is still wrong?

**Comment by Fabrizio-Caruso** [comment]
@oliverschmidt Who are the TGI and ATARI_LYNX experts that could fix this bug?

**Comment by polluks** [comment]
Please take a look
![image_thumb15](https://user-images.githubusercontent.com/4707573/142412220-4ba0a219-8e32-4863-939d-4e40304333f0.png)
https://atarilynxdeveloper.wordpress.com/2012/04/25/programming-tutorial-part-6colors/

**Comment by Fabrizio-Caruso** [comment]
@polluks I know how to display black on white by avoiding some commands but the code above should work. It works with red but not with black,

**Comment by Fabrizio-Caruso** [comment]
@greg-king5, @oliverschmidt could this issue be tagged as "bug" ?
If not why?

**Comment by karrika** [comment]
There is actually only one solution that would make the text work correctly with index 0 (TGI_BLACK).
The solution is to create the characters as shaped sprites. This would add complexity to the text font as some characters need to be split up into two sprites. This is because we cannot draw holes if we don't use index 0 as transparency.
The fix would be possible to implement.

**Comment by Fabrizio-Caruso** [comment]
@mrdudz conio? There is still no CONIO support for this target  (#960). @greg-king5 has made some experiments but there was a problem with some characters.

**Comment by Fabrizio-Caruso** [comment]
@karrika could we have a different index for "real black"? Sorry for the silly question. I understand index 0 is transparent but maybe we can also have normal black color at a different index.

**Comment by mrdudz** [comment]
> @mrdudz conio? There is still no CONIO support for this target (#960). @greg-king5 has made some experiments but there was a problem with some characters.

Sorry, TGI then. Same applies :)

**Comment by karrika** [comment]
> @karrika could we have a different index for "real black"? Sorry for the silly question. I understand index 0 is transparent but maybe we can also have normal black color at a different index.

The TGI specs define that index 0 is the darkest colour and the brightest colour has the highest index.

So my opinion is to change the font to work with BACKGROUND sprites and shaped drawing. In this case the Suzy draws only the pixels of the pen colour and there is no transparency.

Personally I only use a stripped down TGI driver without text functionality. Because then I can use any style in my texts. Like the example of a game I am working on right now.
![Screenshot](https://user-images.githubusercontent.com/6996727/155321839-78dd4279-8ea8-4c7b-8b50-3a758b2d186c.jpg)

**Comment by Fabrizio-Caruso** [comment]
@karrika any technical solution that would work with the same TGI APIs would be ok.

Could such a solution could be used to also  have some CONIO support? https://github.com/cc65/cc65/issues/960

Having support for other fonts would be great. https://github.com/cc65/cc65/issues/1582

Please excuse my question if it makes no sense:
If BLACK has to be indexed as zero, could't we have transparent mapped to a different index outside the range. I guess it is not as simple...

**Comment by mrdudz** [comment]
Pardon my ignorance - but couldnt the TGI color 0 simply use a different hardware color index? (Ideally this would then be consistant across other APIs, however not strictly required either)

**Comment by karrika** [comment]
The transparent has to be index 0. There is no way to change that.

I am currently creating my fonts with a baseline and variable widths for the characters. Perhaps the easiest way to deal with this would be to have several different tgi-drivers. One driver for the current font that confirms to the specs and fixes this bug and one driver for no text.

We could also have some simple tool for creating a custom tgi-driver with the kind of font you want.

**Comment by karrika** [comment]
> Pardon my ignorance - but couldnt the TGI color 0 simply use a different hardware color index? (Ideally this would then be consistant across other APIs, however not strictly required either)

Yes it could. But the default palette was created to follow the Dawnbringer colours. Almost all of my games use that palette. Changing the hardware palette requires the programmer to add a pen palette to every sprites and eats up lots of RAM. Not good imho.

**Comment by mrdudz** [comment]
> But the default palette was created to follow the Dawnbringer colours.

Please elaborate, that tells me nothing :)

**Comment by karrika** [comment]
> > But the default palette was created to follow the Dawnbringer colours.
> 
> Please elaborate, that tells me nothing :)

Lol! The Dawnbringer palette is an optimized 16 colour setup where areas of the colours are close to each other. When you shade something you pick colours left of the colour you are using. If you highlight you pick a colour to the right of the colour you are using. The order of colours is also in groups. Water/sky, skin, plants, earth. So it is not really a technical thing - more an artistic idea.
A girl from Australia optimized the palette for the Lynx and I was so happy with the result that I have been using just the default colours since then.

**Comment by mrdudz** [comment]
OK. And there was i thinking we take the C64 palette as a reference point, like we do with conio =)

**Comment by Fabrizio-Caruso** [comment]
@karrika yes, you are right but can't we reserve an index for transparent and lose one color? Black cannot be sacrificed as it is probably the most commonly used color.
If someone can implement a better solution with no sacrifice, it would be better, though.

**Comment by mrdudz** [comment]
I know nothing about the lynx, so i'll have to take your word for whatever you say :) So did i understand correctly that there are 16 available hardware colors, of which one (index 0) will be transparent? what if we just shift all colors up one index and have 15 colors?

Then, i am not sure rendering (only) text as sprites to get black is really the solution. TGI isnt just for text - so what about all the other things you can do with it? how will those work with black?

**Comment by karrika** [comment]
> I know nothing about the lynx, so i'll have to take your word for whatever you say :) So did i understand correctly that there are 16 available hardware colors, of which one (index 0) will be transparent? what if we just shift all colors up one index and have 15 colors?
> 
> Then, i am not sure rendering (only) text as sprites to get black is really the solution. TGI isnt just for text - so what about all the other things you can do with it? how will those work with black?

That would work.
![Screenshot from 2022-02-23 15-44-53](https://user-images.githubusercontent.com/6996727/155331674-6b43c592-4618-4bd2-9703-e6599a46b08b.png)

Usually I use the 17th colour that does not exist when I design the sprites. This allows me to use all 16 colours for shaped sprites. But if you are happy with a 15 colour palette then I might sacrifice colour index 11 which is DARK_BROWN. And move the colours from index 0 to 10 to become hardware indices 1 to 11.

Be aware! There is no index 0 anymore after this as TGICOLOR_BLACK = 1.

**Comment by karrika** [comment]
I just remembered. Credits of the Lynx palette to AtariAge user @Cammy
She did a great job with the default palette.

**Comment by mrdudz** [comment]
Multiple shades of brown... sure that isnt a C64 palette? :o)

**Comment by Fabrizio-Caruso** [comment]
@mrdudz If I remember well, CC65 tries to approximate the C64 palette in CONIO. TGI colors can have different mappings. So maybe we can leave the current palette except for one color maybe.

**Comment by karrika** [comment]
I did have a look at the sources. My suggestion is to define a new colour TGI_COLOR_TRANSPARENT that is 0

We also need to set the default bgcolor to 0 as this is what people want.

Then we need to set the default drawindex in the driver to 1 as this is black text. Also what people may want.

**Comment by Fabrizio-Caruso** [comment]
@karrika Will the use be able to set a different bicolor? Will everything work with a white background?

**Comment by karrika** [comment]
> @karrika Will the use be able to set a different bicolor? Will everything work with a white background?

The normal way is to paint the paper white with tgi_bar(0, 0, 159, 101).
After that you paint the text with transparent background.
To highligt the text you set the bgcolour to yellow and print black text.
To end the highlight set bgcolor to transparent again,

**Comment by Fabrizio-Caruso** [comment]
@karrika the standard TGI way should just eb        
```tgi_setbgcolor(TGI_COLOR_WHITE);```

Is your proposed change going to break this?
I mean will your propose change allow the use of `tgi_setbgcolor` to set background to anything different from black and then use black or any other color to display text?

Is your change making black special again (in a different way)?
We should have no special color for the APIs. All colors should be treated equally. At least this is how I understand TGI on all other targets.

**Comment by karrika** [comment]
> @karrika the standard TGI way should just be `tgi_setbgcolor(TGI_COLOR_WHITE);`
> 
> Is your proposed change going to break this? I mean will your propose change allow the use of `tgi_setbgcolor` to set background to anything different from black and then use black or any other color to display text?
> 
> Is your change making black special again (in a different way)? We should have no special color for the APIs. All colors should be treated equally. At least this is how I understand TGI on all other targets.

Black is not special in any way. But the bgcolor does not paint the pixels between lines. So if you paint text with bgcolor=white and pen=black you will get 8 pixel high lines with white backgrounds. But the color between the lines may be whatever was on the screen before.

Also the tgi_init will fill the screen with zero I believe. It should be changed to fill the screen with black to keep the screen black after tgi_init().

**Comment by karrika** [comment]
There is no defined y coordinate for the next line in TGI. So of course if you don't leave extra space between the lines things work as you want.

But I often print lines at every 10 pixels.

**Comment by mrdudz** [comment]
I suggest making a small test program (which could well be added into samples later), run it on one or two other targets (like c64 and atari) and make sure it produces equivalent output on all of them

**Comment by oliverschmidt** [comment]
Maybe / likely I'm missing the point, but nevertheless I'd like to point out that TGI_COLOR_BLACK doesn't have to have the value 0. Rather, every target (not every TGI driver!) can define its individual values for TGI_COLOR_BLACK and TGI_COLOR_WHITE.

E.g. the Apple II defines TGI_COLOR_WHITE to be 3 - see https://github.com/cc65/cc65/blob/master/libsrc/apple2/tgi_colors.s

**Comment by polluks** [comment]
> I am currently creating my fonts with a baseline and variable widths for the characters. Perhaps the easiest way to deal with this would be to have several different tgi-drivers. One driver for the current font that confirms to the specs and fixes this bug and one driver for no text.

Better use tgi_ioctl() for changing a font, it's less overhead.

**Comment by karrika** [comment]
> > I am currently creating my fonts with a baseline and variable widths for the characters. Perhaps the easiest way to deal with this would be to have several different tgi-drivers. One driver for the current font that confirms to the specs and fixes this bug and one driver for no text.
> 
> Better use tgi_ioctl() for changing a font, it's less overhead.

I am not so sure about that. ROM space is cheap compared to RAM. I rather have several copies of things in ROM than any extra stuff that clutters up RAM. Currently Wizzy's ROM space is 2MB while the RAM is still at 48k.

It is simple to create your own fonts and use someting like tgi_arialxy(x, y, msg) for printing something with a new font.

In games you often just need a subset of characters like just the numbers 0..9
Or you may want to add some emoticons, musical symbols and map them to unused ASCII characters.

So if you really want to extend the tgi to use different fonts I would go for a system where you can insert a pointer to your own font in the tgi driver. Then you can optimize the font to have only the stuff you need for the game.

Perhaps tgi_ioctl(nr, pointer to outtext routine for your own font)?

**Comment by karrika** [comment]
I just forked the repo and ran a test to see how broken it is.
```
include <tgi.h>
#include <6502.h>
#include <lynx.h>

void main()
{
        tgi_install(&tgi_static_stddrv);
        tgi_init();
        CLI();
        while (tgi_busy())
                ;
        // Draw paper white top black bottom
        tgi_setcolor(COLOR_WHITE);
        tgi_bar(0,0,159,51);
        tgi_setcolor(COLOR_BLACK);
        tgi_bar(0,52,159,101);

        // Draw black text on white paper
        tgi_setcolor(COLOR_BLACK);
        tgi_outtextxy(5,10,"Black text");

        // Draw white text on black paper
        tgi_setcolor(COLOR_WHITE);
        tgi_outtextxy(5,10 + 52,"White text");

        // Draw white text with black bg on white paper
        tgi_setbgcolor(COLOR_BLACK);
        tgi_setcolor(COLOR_WHITE);
        tgi_outtextxy(5,20,"White text black bg");

        // Draw black text with white bg on white paper
        tgi_setbgcolor(COLOR_WHITE);
        tgi_setcolor(COLOR_BLACK);
        tgi_outtextxy(5,20 + 52,"Black text white bg");

        tgi_updatedisplay();
}
```
![Screenshot from 2022-02-24 08-05-27](https://user-images.githubusercontent.com/6996727/155467775-c6178246-83a1-4ebd-a074-ad276e4bc70e.png)

Everything works except black text on white canvas.

**Comment by karrika** [comment]
I did some changes to the repo and will make a pull request. At least the result is as expected for this case.
![Screenshot from 2022-02-24 08-25-48](https://user-images.githubusercontent.com/6996727/155470132-41a1a206-f348-4045-9bfb-c6de90ecbf25.png)

**Comment by karrika** [comment]
A typical small font for the Lynx contains just sprites that have their anchor point on the baseline. So there is no need to maintain a baseline in the code.

A character could look like this:
```
#define one_BPP          4
#define one_WIDTH        3
#define one_HEIGHT       6
static const unsigned char one[] = {
    0x04,0x90,0x04,0x00,0x01,0x04,0x90,0x04,0x00,0x04,0x90,0x04,0x00,0x04,0x90,0x04,
    0x00,0x04,0x90,0x04,0x00,0x04,0x90,0x44,0x00,0x00,
};

```
And a font like this:
```
static char *table[] = {
        zero,
        one,
        two,
        three,
        four,
        five,
        six,
        seven,
        eight,
        nine
};

static char *number(char ch)
{
        return table[ch - '0'];
}

```

**Comment by karrika** [comment]
If you are interested in making fonts for the Lynx the sp65 has some nice options for that.

Makefile:
```
buildfont: statustiles.pcx
        $(MAKE) addchar SLICE=0,96,3,7 CHAR=one BASELINE=6
        $(MAKE) addchar SLICE=8,96,6,7 CHAR=two BASELINE=6
        $(MAKE) addchar SLICE=16,96,6,7 CHAR=three BASELINE=6

addchar:
        $(SP) -r statustiles.pcx --slice $(SLICE) -c lynx-sprite,mode=literal,ax=0,ay=$(BASELINE),pen=08 \
                -w $(CHAR).c,ident=$(CHAR),bytesperline=8

```

The special designators (ax,ay) define the anchor point of the character. The pixel from where the drawing starts.
The pen=08 defines that it is a binary sprite with transparent and colour 8 (yellow in my sprite sheet).
This allows me to keep my graphics in a multi-coloured sheet and I can extract the font as a BPP_1 to save space.

**Comment by Fabrizio-Caruso** [comment]
I would also like to define 8x8 non-transparent 8x8 (or 7x8 if simpler) monochromatic pseudo-tiles (just objects to be displayed at 8x8 boundaries. If possible I would like to keep their definitions in the code sort of like as you propose with the char arrays. Could you please provide an example on how to display such an object? Could we please have a thread on the discussion here on GitHub?

**Comment by karrika** [comment]
Sure. I already wrote a small animated example of tiles I found in your repo.
Perhaps we don't clutter this thread.
[literaldemo.zip](https://github.com/cc65/cc65/files/8134269/literaldemo.zip)

**Comment by karrika** [comment]
This issue was solved by pull request #1686 
It could be closed now.
